File: agent/sensors/bpf_syscall.c
// SPDX-License-Identifier: Apache-2.0
// eBPF program: monitor bpf() syscall invocations
//
// Attaches to tracepoint/syscalls/sys_enter_bpf and captures
// BPF_PROG_LOAD, BPF_PROG_ATTACH, and BPF_MAP_CREATE commands.

//go:build ignore

#include "vmlinux.h"
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>
#include <bpf/bpf_core_read.h>

#define TASK_COMM_LEN 16
#define BPF_PROG_NAME_LEN 16

#define BPF_MAP_CREATE    0
#define BPF_PROG_LOAD     5
#define BPF_PROG_ATTACH   8

#define EVENT_BPF_LOAD 1

struct hook_event {
    u32 event_type;
    u32 pid;
    u32 uid;
    u32 gid;
    u32 ppid;
    char comm[TASK_COMM_LEN];
    u32 bpf_cmd;
    u32 prog_type;
    char prog_name[BPF_PROG_NAME_LEN];
    u32 attach_type;
    u32 insn_count;
    u64 insns_ptr;   // userspace pointer to BPF instructions (for hash computation)
};

struct {
    __uint(type, BPF_MAP_TYPE_RINGBUF);
    __uint(max_entries, 256 * 1024);
} events SEC(".maps");

SEC("tracepoint/syscalls/sys_enter_bpf")
int trace_bpf_enter(struct trace_event_raw_sys_enter *ctx)
{
    u32 cmd = (u32)ctx->args[0];

    // Only capture program load, attach, and map create
    if (cmd != BPF_PROG_LOAD && cmd != BPF_PROG_ATTACH && cmd != BPF_MAP_CREATE)
        return 0;

    struct hook_event *e;
    e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
    if (!e)
        return 0;

    e->event_type = EVENT_BPF_LOAD;
    e->bpf_cmd = cmd;

    u64 pid_tgid = bpf_get_current_pid_tgid();
    e->pid = pid_tgid >> 32;

    u64 uid_gid = bpf_get_current_uid_gid();
    e->uid = (u32)uid_gid;
    e->gid = uid_gid >> 32;

    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    e->ppid = BPF_CORE_READ(task, real_parent, tgid);

    bpf_get_current_comm(&e->comm, sizeof(e->comm));

    // Read bpf_attr for program details (only for PROG_LOAD)
    if (cmd == BPF_PROG_LOAD) {
        union bpf_attr *attr = (union bpf_attr *)ctx->args[1];
        bpf_probe_read_user(&e->prog_type, sizeof(e->prog_type), &attr->prog_type);
        bpf_probe_read_user_str(e->prog_name, sizeof(e->prog_name), (void *)attr->prog_name);
        bpf_probe_read_user(&e->insn_count, sizeof(e->insn_count), &attr->insn_cnt);

        // Capture the userspace pointer to BPF instructions for hash computation
        u64 insns;
        bpf_probe_read_user(&insns, sizeof(insns), &attr->insns);
        e->insns_ptr = insns;
    }

    if (cmd == BPF_PROG_ATTACH) {
        union bpf_attr *attr = (union bpf_attr *)ctx->args[1];
        bpf_probe_read_user(&e->attach_type, sizeof(e->attach_type), &attr->attach_type);
    }

    bpf_ringbuf_submit(e, 0);
    return 0;
}

char LICENSE[] SEC("license") = "Dual BSD/GPL";

File: agent/sensors/bpf_syscall.go
//go:build linux

package sensors

import (
	"bytes"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"fmt"
	"os"
	"time"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"
	"github.com/cilium/ebpf/ringbuf"
	"github.com/google/uuid"

	"github.com/dlenrow/hookmon/pkg/event"
)

// bpfSyscallEvent mirrors the C struct hook_event in bpf_syscall.c.
// The C struct uses __attribute__((packed)) to match Go's binary.Read layout.
type bpfSyscallEvent struct {
	EventType  uint32
	PID        uint32
	UID        uint32
	GID        uint32
	PPID       uint32
	Comm       [16]byte
	BPFCmd     uint32
	ProgType   uint32
	ProgName   [16]byte
	AttachType uint32
	InsnCount  uint32
	InsnsPtr   uint64
}

// BPFSyscallSensor monitors the bpf() syscall for program loading and attachment.
type BPFSyscallSensor struct {
	eventCh chan *event.HookEvent
	tp      link.Link
	reader  *ringbuf.Reader
	coll    *ebpf.Collection
	done    chan struct{}
}

// NewBPFSyscallSensor creates a new bpf() syscall monitor sensor.
func NewBPFSyscallSensor() *BPFSyscallSensor {
	return &BPFSyscallSensor{
		eventCh: make(chan *event.HookEvent, 256),
		done:    make(chan struct{}),
	}
}

func (s *BPFSyscallSensor) Name() string       { return "bpf_syscall" }
func (s *BPFSyscallSensor) Type() SensorType   { return SensorTypeBPF }

func (s *BPFSyscallSensor) Start() error {
	spec, err := ebpf.LoadCollectionSpecFromReader(bytes.NewReader(bpfSyscallBPF))
	if err != nil {
		return fmt.Errorf("load BPF spec: %w", err)
	}

	s.coll, err = ebpf.NewCollection(spec)
	if err != nil {
		return fmt.Errorf("create BPF collection: %w", err)
	}

	prog := s.coll.Programs["trace_bpf_enter"]
	if prog == nil {
		return errors.New("BPF program 'trace_bpf_enter' not found in collection")
	}

	s.tp, err = link.Tracepoint("syscalls", "sys_enter_bpf", prog, nil)
	if err != nil {
		return fmt.Errorf("attach tracepoint: %w", err)
	}

	eventsMap := s.coll.Maps["events"]
	if eventsMap == nil {
		return errors.New("BPF map 'events' not found in collection")
	}

	s.reader, err = ringbuf.NewReader(eventsMap)
	if err != nil {
		return fmt.Errorf("create ringbuf reader: %w", err)
	}

	go s.readLoop()
	return nil
}

func (s *BPFSyscallSensor) Stop() error {
	close(s.done)
	if s.reader != nil {
		s.reader.Close()
	}
	if s.tp != nil {
		s.tp.Close()
	}
	if s.coll != nil {
		s.coll.Close()
	}
	return nil
}

func (s *BPFSyscallSensor) Events() <-chan *event.HookEvent { return s.eventCh }

func (s *BPFSyscallSensor) readLoop() {
	for {
		record, err := s.reader.Read()
		if err != nil {
			select {
			case <-s.done:
				return
			default:
				continue
			}
		}

		var raw bpfSyscallEvent
		if err := binary.Read(bytes.NewReader(record.RawSample), binary.LittleEndian, &raw); err != nil {
			continue
		}

		evtType := event.EventBPFLoad
		if raw.BPFCmd == 8 { // BPF_PROG_ATTACH
			evtType = event.EventBPFAttach
		}

		// Compute BPF bytecode hash if we have the instructions pointer
		progHash := ""
		if raw.BPFCmd == 5 { // BPF_PROG_LOAD
			if raw.InsnsPtr != 0 && raw.InsnCount > 0 {
				progHash = computeProgHash(raw.PID, raw.InsnsPtr, raw.InsnCount)
			}
		}

		hookEvt := &event.HookEvent{
			ID:        uuid.New().String(),
			Timestamp: time.Now(),
			EventType: evtType,
			Severity:  event.SeverityWarn,
			PID:       raw.PID,
			PPID:      raw.PPID,
			UID:       raw.UID,
			GID:       raw.GID,
			Comm:      nullTermStr(raw.Comm[:]),
			BPFDetail: &event.BPFDetail{
				BPFCommand: raw.BPFCmd,
				ProgType:   raw.ProgType,
				ProgName:   nullTermStr(raw.ProgName[:]),
				AttachType: raw.AttachType,
				InsnCount:  raw.InsnCount,
				ProgHash:   progHash,
			},
		}

		select {
		case s.eventCh <- hookEvt:
		default:
			// Drop event if channel full
		}
	}
}

// nullTermStr converts a null-terminated byte slice to a string.
func nullTermStr(b []byte) string {
	n := bytes.IndexByte(b, 0)
	if n < 0 {
		n = len(b)
	}
	return string(b[:n])
}

// computeProgHash reads BPF bytecode from /proc/<pid>/mem and computes its SHA256.
// Each BPF instruction is 8 bytes, so total size = insn_count * 8.
// This works because the tracepoint fires during the bpf() syscall while the
// process is still alive and the instructions buffer is still in its address space.
func computeProgHash(pid uint32, insnsPtr uint64, insnCount uint32) string {
	const bpfInsnSize = 8
	size := int64(insnCount) * bpfInsnSize
	if size <= 0 || size > 1<<20 { // sanity: cap at 1MB
		fmt.Fprintf(os.Stderr, "computeProgHash: bad size %d (insn_count=%d)\n", size, insnCount)
		return ""
	}

	memPath := fmt.Sprintf("/proc/%d/mem", pid)
	f, err := os.Open(memPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "computeProgHash: open %s: %v\n", memPath, err)
		return ""
	}
	defer f.Close()

	buf := make([]byte, size)
	n, err := f.ReadAt(buf, int64(insnsPtr))
	if err != nil {
		fmt.Fprintf(os.Stderr, "computeProgHash: readat offset=0x%x size=%d: %v (read %d bytes)\n",
			insnsPtr, size, err, n)
		return ""
	}

	h := sha256.New()
	h.Write(buf[:n])
	hash := fmt.Sprintf("sha256:%x", h.Sum(nil))
	fmt.Fprintf(os.Stderr, "computeProgHash: pid=%d insns=0x%x count=%d hash=%s\n",
		pid, insnsPtr, insnCount, hash)
	return hash
}

// bpfSyscallBPF is provided via go:embed in embed.go
